Add minimal version of in repository mandatory feature flags.

textshell opened the pull request:
This should allow us to make sure older borg versions can be cleanly
prevented from doing operations that are no longer safe because of
repository format evolution. This allows more fine grained control than
just incrementing the manifest version. So for example a change that
still allows new archives to be created but would corrupt the repository
when an old version tries to delete an archive or check the repository
would add the new feature to the check and delete set but leave it out
of the write set.

This is somewhat inspired by ext{2,3,4} which uses sets for
compat (everything except fsck), ro-compat (may only be accessed
read-only by older versions) and features (refuse all access).

This implements #1806

Open questions:

- [x] What granularity is right? I currently use "write" for create, "delete" for delete and prune and "check" for everything else.
- [x] Changes that make delete unsafe likely have consequences for the cache. Should this be handled or should we just wait until the first feature that need this and think about it then?
- [x] Should this also add support for manifest version 2 (just manifest version 1 plus respects the mandatory feature flags), so we have a way to lockout 1.0.x borg versions for changes where this patch already makes it safe with borg 1.1.x?

ThomasWaldmann commented:
granularity: as you said + read, for completeness.
delete / cache: no idea
manifest v2: not sure how this helps / how you'ld use it.

textshell commented:
ok, i'll add a read one too. We could add a second one for reading into the archives. Shouldn‘t be much work.
The basic idea about manifest v2 is, that all these mandatory feature flags won‘t stop borg 1.0.9 from possibly corrupting a repo where borg 1.1 would correctly refuse to do disallowed actions. So if we want to prevent older versions we need to change something they check once to make sure mandatory feature support is there. One easy way would be to use the manifest version. We can still do that later of course. I tend to accepting v2 as soon as this is in. (just changing `if not m.get(b'version') == 1:` to `if m.get(b'version') not in (1,2):` in Manifest.load)

textshell commented:
The cache problem is as follows. Say we want to add a feature that allows larger archives by some means where not all chunk id‘s containing items are stored in the archive chunk directly.
Such a repository would still be safe for adding archives to with an old version. So such a feature should only be mandatory for "delete" and "check". But of course older borg versions would not have the chunks in a large archive be refcounted correctly in the local cache. It i was wondering if we need to store some additional flags with the cache so borg versions can coordinate not to corrupt things. I never much thought about cache changes though.

ThomasWaldmann commented on borg/archiver.py:
```diff
@@ -177,6 +178,7 @@ def do_check(self, args, repository):
     @with_repository()
     def do_change_passphrase(self, args, repository, manifest, key):
         """Change repository key file passphrase"""
+        check_repository_compatibility(manifest, REPO_OPERATION_CHECK)
```
having to add that everywhere to avoid damage seems error prone, it could be forgotten.
so, add it to the decorator maybe?

textshell replied:
I was unsure how to make that nice.
- There are places where the manifest is loaded only later. So it can not really be universal
- should this be additional in the exiting decorators or would it be stacked?
- I really want this to be in 1.1, so i do consider it ok if it needs to be optimized later.

Maybe we can just load the manifest more often centrally that would certainly make this easier as decorator?

ThomasWaldmann replied:
true, the decorator is not always used. but it would cover at least the places where it is used.

additional in the existing decorator or just an additional call as you have now.

textshell replied:
I made a version using a `compatibility` argument to `with_repository`
Do you like this version better? I made it very explicit about locations where there is no checking.

ThomasWaldmann commented on borg/helpers.py:
```diff
@@ -98,6 +101,23 @@ def check_extension_modules():
+    if manifest is None:
+        # various fake options can have manifest is None
+        return
+    if not operation in manifest.config:
```
"not in" is a operator.

ThomasWaldmann commented on borg/helpers.py:
```diff
@@ -98,6 +101,23 @@ def check_extension_modules():
+        return
+    requirements = manifest.config[operation]
+    if 'mandatory' in requirements and not set(requirements['mandatory']) <= SUPPORTED_REPO_FEATURES:
+        raise MandatoryFeatureUnsupported(set(requirements['mandatory']) - SUPPORTED_REPO_FEATURES)
```
maybe always compute the unsupported_features set and raise if not empty? saves a bit of duplication maybe.

ThomasWaldmann commented on borg/archiver.py:
```diff
@@ -54,7 +56,8 @@ def argument(args, str_or_bool):
 
-def with_repository(fake=False, invert_fake=False, create=False, lock=True, exclusive=False, manifest=True, cache=False):
+def with_repository(fake=False, invert_fake=False, create=False, lock=True,
+                    exclusive=False, manifest=True, cache=False, compatibility=None):
```
compatibility is missing in the docstring below.

ThomasWaldmann commented on borg/archiver.py:
```diff
@@ -69,6 +72,11 @@ def with_repository(fake=False, invert_fake=False, create=False, lock=True, excl
+            if create:
+                compatibility = REPO_OPERATION_NO_CHECK
+            if compatibility is None:
+                raise Exception("with_repository called without compatibility argument")
```
the call to check_repository_compatibility() is still missing here (somewhere below), right?

textshell replied:
Yes, i’ll fix that in the next revision.

enkore replied:
maybe put that into Manifest.load, not here?

enkore commented:
Thoughts

- What is READ actually for? Wouldn't we indicate this rather via version attributes on objects (manifest, archives)?
  - Some of these are not super helpful (`assert` form), but I'd rather fix these and give them better error messages
  - Perhaps to indicate a feature present in the repo that is required to be understood for a cache sync? In this case it would be complimentary to the versions, to make it fail fast, but version attributes would still provide the safety called for.
- WRITE -> CREATE?
  - Does WRITE/CREATE imply READ?
    - Right now it does, but one doesn't need a lot of imagination to see that change: eg. repository-provided bloom filter

enkore commented on borg/helpers.py:
```diff
@@ -98,6 +102,28 @@ def check_extension_modules():
+        raise MandatoryFeatureUnsupported(set(requirements['mandatory']) - SUPPORTED_REPO_FEATURES)
+
+
 class Manifest:
```
v `def load(cls, repository, key=None, operation=None, force_tam_not_required=False):`

enkore commented:
If operation levels don't have a total ordering then it might make more sense to be able to verify legal operation against a set of operation levels than just one.

ThomasWaldmann commented:
if this is going to still take a while: shift it to b5?

textshell commented:
Sorry this got stuck a bit.
I now moved the compatibility checking into the manifest class. Still the majority of the checking is specified by the with_repository decorator, but special cases now use Manifest.load for further checking.

About READ: I originally left read out because i felt that likely something at archive level would be more helpful. So i’m happy with or without it here. I do like the idea to have a REFCOUNT_SYNC operation quite a lot. Maybe instead of READ?

WRITE currently only covers do_create, so maybe we should rename it. In a way WRITE does not really need perfect REFCOUNT_SYNC, it just needs a conservative refcount that doesn‘t have false positives.

Checking more than one might be needed in some situations, esp. if we go with REFCOUNT_SYNC and for now just keep do_create simple by requiring both.

enkore commented:
> WRITE currently only covers do_create, so maybe we should rename it. In a way WRITE does not really need perfect REFCOUNT_SYNC, it just needs a conservative refcount that doesn‘t have false positives.
>
> Checking more than one might be needed in some situations, esp. if we go with REFCOUNT_SYNC and for now just keep do_create simple by requiring both.

Exactly my thoughts :)

ThomasWaldmann commented on borg/helpers.py:
```diff
@@ -146,8 +159,18 @@ def load(cls, repository, key=None, force_tam_not_required=False):
+            return
+        requirements = self.config[operation]
+        if 'mandatory' in requirements and not set(requirements['mandatory']) <= SUPPORTED_REPO_FEATURES:
+            raise MandatoryFeatureUnsupported(set(requirements['mandatory']) - SUPPORTED_REPO_FEATURES)
```
see https://github.com/borgbackup/borg/pull/2134#discussion_r100644624

textshell replied:
fixed.

textshell commented on borg/helpers.py:
```diff
@@ -98,6 +102,15 @@ def check_extension_modules():
+REPO_OPERATION_CHECK = 'check'
+REPO_OPERATION_WRITE = 'write'
+REPO_OPERATION_DELETE = 'delete'
+REPO_OPERATION_NO_CHECK = '-'
```
Yes, check and no_check seem ambiguous, that needs improving. Replace REPO_OPERATION by OP_ and move in as members of Manifest?

enkore replied:
Make it a faux enum instead?

```py
class Manifest:
  # possibly as a class of manifest
  class Operation:
    READ = '...
    WRITE = '...
```

textshell replied:
Now implemented with the stdlib enum class/metaclass. I think requiring 3.4 is ok? Or do we still target 3.3?

textshell commented on borg/archiver.py:
```diff
@@ -65,7 +68,18 @@ def with_repository(fake=False, invert_fake=False, create=False, lock=True, excl
+
+    if not create and (manifest or cache):
+        if compatibility is None:
+            raise Exception("with_repository called without compatibility argument")
```
The way this is currently written, it raises these on first import of archiver.py. So these are not really exceptions that might need to be caught but more sanity checks done while loading the module. So i don‘t think this matters much.

enkore replied:
The assertion in the else branch is sufficient for both cases; just move the description there?

enkore commented on borg/helpers.py:
```diff
@@ -146,8 +159,18 @@ def load(cls, repository, key=None, force_tam_not_required=False):
+            return
+        if operation not in self.config:
+            return
+        requirements = self.config[operation]
```
~ maybe use a separate dict in `config` for this, like `feature_flags` or so, to keep the `config` namespace clean.

textshell replied:
currently it is structured
`manifest.config[b'write'] = {b'mandatory': [b'unknown-feature']}
`
But i’m open for other layouts. I think there are some features we want to store in the config that don‘t lockout older versions. So i would want to keep some distinction between mandatory and "as supported" features.

textshell commented on borg/helpers.py:
```diff
@@ -86,6 +86,10 @@ class PlaceholderError(Error):
 
 
+class MandatoryFeatureUnsupported(Error):
+    """Unsupported repository feature(s) {}. A newer version is required to access this repository."""
```
now "a newer version of borg"

textshell commented:
I hope i fixed all the clear cut review comments now. I decided to just include support for manifest version 2, worst case we wasted a version number.

Still open is encore’s question what `READ` actually means and the idea to have a dedicated `REFCOUNT_SYNC`. `REFCOUNT_SYNC` has overlap with `DELETE` because delete always needs correct refcounts.

I’m tempted by a fallback that still allows archive creation without `REFCOUNT_SYNC` but that would be hard to test unless we offered it as a genuine operation mode outside of some obscure fallback code path. `REFCOUNT_SYNC` would solve cache compatibility because that is the only critical cached data that needs handling at the moment it think.

enkore commented:
It's a bit of a coincidence, since I just finished [a prototype](https://gist.github.com/enkore/fce238674acc3963ae1b1f00be8c08a6) that does exactly what REFCOUNT_SYNC would be needed for, i.e. `borg create` without a cache (sync).

If we separate `REFCOUNT_SYNC` from `READ` then yeah... what is `READ` for?

textshell commented:
From discussion on irc:

For cache maintenance the basic principles are:
Wipe when
* the client understands a new features that has previously been ignored, OR
* a client doesn't understand a feature that was previously understood.

This can be implement fairly easily:
Each client version has a set of features which contains features that the client knows to handle `client_features`
When accessing a repository collect all mandatory features from the manifest into a `flat_featues` set.
Save in the cache two sets called `ignored_features` and `mandatory_features`.
When updating the cache also update:
`ignored_features` += `flat_featues` - `client_features`
`mandatory_features` += `flat_featues` & `client_features`

when accessing the cache:
if `ignored_features` contains any item from `client_features`: wipe and rebuild cache
if `mandatory_features` contains any item not in `client_features`: wipe and rebuild cache

Possible optimization. A future client could tweak the rules to in cases where it is known to be save in order to reduce rebuilds. But that will always be a case by case decision that can only be made when a new feature will be added, so this PR does not need to consider this in detail.

textshell commented:
i rebased this to something more recent and implemented the cache wiping strategy from the comment above. This should solve the cache problem.
Possibly we might want skip some of the operations when deciding what feature flags to use for cache compatibility. I did not do this in this patch. But possibly ignoring write would make some sense. This is only useful if we expect features that are mandatory for write but not for delete or check. One kind of feature where this might be useful would be a feature that just prevents regressions in structure after some kind of whole repository upgrades.

I lightly tested this manually. It still needs automated tests. One thing i'm unsure about is how to best test that in all the existing tests the cache wipe is not triggered. That would need some rather global way to make the cache wipe method throw an exception. Maybe a global test fixture would do, but i'm unsure how to disable that in just some tests.

Also still pending is the question what to do with the "READ" operation:
* Move to archive
* Remove
* Keep as is
I thing the first option would be best with a super simple model where each archive can have a feature_flags list of strings that are all mandatory. Not sure how mount of the whole repo would handle this though.

enkore commented on borg/archiver.py:
```diff
@@ -537,7 +551,7 @@ def do_delete(self, args, repository):
                 msg = []
                 try:
-                    manifest, key = Manifest.load(repository)
+                    manifest, key = Manifest.load(repository, Manifest.NO_OPERATION_CHECK)
```
minor;

given that every(?) time Manifest.load is called explicitly (outside with_repository) NO_OPERATION_CHECK is passed, make it the default?

enkore replied:
I meant the default on `Manifest.load` itself, though. Fair enough either way, you make a solid point.

enkore commented on borg/archiver.py:
```diff
@@ -67,7 +68,20 @@ def with_repository(fake=False, invert_fake=False, create=False, lock=True, excl
+        if compatibility is not None:
+            raise Exception("with_repository called with compatibility argument but would not check" + repr(compatibility))
+        if create:
+            compatibility = Manifest.NO_OPERATION_CHECK
```
? do_create below specifies compatibility?

enkore commented on borg/helpers.py:
```diff
@@ -173,8 +192,34 @@ def load(cls, repository, key=None, force_tam_not_required=False):
+            if feature_flags is None:
+                return
+            if operation.value.encode() not in feature_flags:
+                return
```
Shouldn't this be `continue`, so the other operations are still checked?

enkore commented on borg/helpers.py:
```diff
@@ -124,8 +129,22 @@ def check_extension_modules():
+
+    NO_OPERATION_CHECK = tuple()
+
+    ALL_POSSIBLE_FEATURES = set([])
```
minor; also a frozenset()?

enkore commented on borg/helpers.py:
```diff
@@ -124,8 +129,22 @@ def check_extension_modules():
         raise ExtensionModuleError
 
 
+SUPPORTED_REPO_FEATURES = frozenset()
```
minor; move this constant to Manifest as well, where the other related constants live as well?

enkore commented on borg/cache.py:
```diff
@@ -390,6 +400,39 @@ def legacy_cleanup():
+        with open(os.path.join(self.path, 'files'), 'wb') as fd:
+            pass  # empty file
+        self.manifest_id = ''
+        self.config.set('cache', 'manifest', '')
```
Reuse existing Cache.delete?

enkore commented:
> I lightly tested this manually. It still needs automated tests. One thing i'm unsure about is how to best test that in all the existing tests the cache wipe is not triggered. That would need some rather global way to make the cache wipe method throw an exception. Maybe a global test fixture would do, but i'm unsure how to disable that in just some tests.

You could patch (unittest.mock.patch) the delete/create method with a mock, then assert it was not called (`not mock.called`).

enkore commented:
> Also still pending is the question what to do with the "READ" operation:
>
>    Move to archive
>    Remove
>    Keep as is
>
> I thing the first option would be best with a super simple model where each archive can have a feature_flags list of strings that are all mandatory. Not sure how mount of the whole repo would handle this though.

I'm not a huge fan of fine-grained "may work / may not work" TBH. Keeping it repository-wide seems easiest to me and has clear semantics; either a version can read the repository, and then it can read all archives (to at least some extent), or it doesn't and it can't read anything.

textshell commented:
@ docs:
Yes. This needs docs. Maybe even in the internal part of the rtd docs.

@ read
I'm a bit torn, just having one large archive (with new structure) doesn't need to mean that other archives are unreadable. But of course one simple global flag is easier and sufficient for just avoiding harm.

enkore commented:
Regarding flag granularity,
it's not just easier from a developer perspective, but like I mentioned above — it's simpler for using it as well; questions such as "Can I restore that backup/archive?" should be easy and confident to answer.

enkore commented:
Code comments; that's probably the begin/submit review flow offered by Github. It seems to always create new threads, even when replying to another comment. It also makes it easy to not publish comments; I don't use it. The approve/reject/global-comment functionality is accessible without using the other parts of that workflow.

enkore commented:
Regarding cache flags,
the cache is composed of a few distinct structures (chunks, files and chunks.archive.d). What do you think about giving a distinct set of flags to each of these? For example, we know that chunks.archive.d wastes disk spaces due to its format; or that the files cache format could be changed to better accommodate cache-sync-less operation.
However, in all of these cases backwards-compatible approaches are clearly viable as well. So I'm thinking leave it as is; having too many joints (version fields, feature flags in different places) just increases the chance that they're stuck and thus pointless.

enkore commented on borg/cache.py:
```diff
@@ -390,6 +400,39 @@ def legacy_cleanup():
+            repo_features.update(features)
+
+        self.ignored_features.update(repo_features - my_features)
+        self.mandatory_features.update(repo_features & my_features)
```
Aha, so it turns out I read this kinda wrong :)

So this is not a per se separate set of flags for the cache. Rather, it checks if the set of flags stored in the config is a subset of the supported repository flags.

enkore replied:
Aha. That makes a lot of sense there. This allows to build a cache that might only reflect some subset of the data in the repository, and then rule operations out that would require a complete cache (say, *delete*).

enkore commented:
To reduce the overall required work (and because this PR is open since before I made the new branching model, I think) I propose to merge this PR into 1.0-maint (as an exception) and cherry pick into master from there.

enkore commented:
Regarding https://github.com/borgbackup/borg/pull/2134#issuecomment-285946168

So, WRITE does not include synchronizing refcounts. WRITE just means you can update the manifest. If/how you get a list of chunks existing in the repository is NOT covered by WRITE. This would basically make it the usually-lowest level of access.

READ means being able to read ALL archives in the repository, but maybe not every piece of data. Say, we add a new item key with some crazy new FS flags — that would not be a mandatory READ feature. But using new kind of crypto, that would be a mandatory READ feature.

DELETE means being able to establish correct reference counts. Which probably means READ + all additions to referencing chunks (think #1681).

CHECK means being able to do all that DELETE does, but also (?).

We don't need answers every yet. Just a hunch that the category is correct. Which looks good to me so far.

enkore commented:
Docs are in #2602. You can merge it into this or we keep it independent, doesn't really matter.

enkore commented:
(?) Idea: Store another key under *config* in the manifest, which gives a short description for each flag. Error messages could then include the descriptions for even better error messages.

enkore commented on borg/cache.py:
```diff
@@ -390,6 +400,39 @@ def legacy_cleanup():
+    def check_cache_compatibility(self):
+        my_features = Manifest.SUPPORTED_REPO_FEATURES
+        if self.ignored_features & my_features:
+            return False
```
I'm not really sure if I understand this correctly.

ignored_features is the set of features that are mandated for some operations, but not those the cache was originally built for, and these were not supported by the client building / last updating the cache.

So `ignored_features & my_features` would be the set of features that were *not* supported by the previous client. But why wipe the cache here?

enkore replied:
So per https://github.com/borgbackup/borg/pull/2134#issuecomment-303220459 

> So ignored_features & my_features would be the set of features that were not supported by the previous client. But why wipe the cache in this instance?

I.e. the repository has had a feature enabled, which the client just now learned, so as per ^ the cache might need rebuilding. But as per ^^ this can be made a case-to-case decision anyway. Ok. Now I get it :)

ThomasWaldmann replied:
would be cool if the hard to understand stuff ends up in some helpful comments in the src. :)

enkore replied:
Long-form docs are in #2602. Doesn't replace a few well-placed comments, though :)

textshell commented:
WRITE currently tries to do a normal chunk cache sync. Which i think is ok to try. Because for write it is always safe to assume certain chunks are not in the repository when they in fact are. The cache sync is currently not robust with regard to strange and new archive formats. That is strictly speaking not a correctness problem, but an unsightly "crash".

CHECK is the highest level, assuming total understanding of everything because trying to ensure internal consistency might corrupt things. DELETE in contrast does not change anything valuable, except that it needs to know about all chunks to free and needs a chunk refcount cache that is 100% accurate.

textshell commented:
Do we still consider this to be a candidate for 1.0-maint? I think it would be nice to have and this branch still targets 1.0-maint as by previous policy.

textshell commented:
To add about "CHECK". This could also cover various parts that are not worth having separate flags for. E.g. backwards compatible changes in the repokey and various other "special" commands that are too rarely used to be worrying about more precise compatibility checks.

ThomasWaldmann commented:
flake8 is unhappy:
```
./conftest.py:45:9: E306 expected 1 blank line before a nested definition, found 0
./conftest.py:57:35: W292 no newline at end of file
./borg/testsuite/archiver.py:1017:27: E712 comparison to True should be 'if cond is True:' or 'if cond:'
./borg/testsuite/archiver.py:1020:9: E303 too many blank lines (2)
./borg/testsuite/archiver.py:1028:5: E303 too many blank lines (2)
```

textshell commented:
I have squashed this changeset down to something more managable. Until e04dad5 this is (modulo fixes for flake8) without any new changes.
The two follow up commits add some additional comments.

I have not yet removed "TMP: debug tune" but i think even if we want to have something like this, it should go in via a separate PR, so i will remove this before merging.

textshell commented:
Ok, i squashed the last fixups and removed some unneeded print in the tests.

@ThomasWaldmann What do you think?

ThomasWaldmann commented:
ok. \o/

as discussed on irc we'll first release this via master in a 1.1.0 beta.
depending on how that works out, it could be in 1.0.11 or 1.0.12 later.

enkore commented:
Who picks this into master?

textshell commented:
I'm currently trying to rebase this to master.

enkore commented:
Reroll may be easier / quicker.
